File: src/com/easy/ClimbingStairs_70.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class ClimbingStairs_70 {

    // Use Fibonacci number concept
    // Time Complexity: O(n)
    // Space complexity: O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 31.8 MB, less than 100.00%
    public int climbStairs(int n) {
        if (n <= 2) {
            return n;
        }
        int fibNminus2 = 1, fibNminus1 = 2;
        for (int i = 2; i < n; i++) {
            int fibN = fibNminus2 + fibNminus1;
            fibNminus2 = fibNminus1;
            fibNminus1 = fibN;
        }

        return fibNminus1;
    }

    // Use Recursive with memorize
    // Time Complexity: O(n)
    // Space complexity: O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 33 MB, less than 5.26%
    public int climbStairs2(int n) {
        int[] mDP = new int[n + 1];

        return helper(0, n, mDP);
    }

    private int helper(int i, int n, int mDP[]) {
        if (i > n) {
            return 0;
        }
        if (i == n) {
            return 1;
        }
        if (mDP[i] != 0) {
            return mDP[i];
        }
        mDP[i] = helper(i + 1, n, mDP) + helper(i + 2, n, mDP);
        return mDP[i];
    }

    // Use DP
    // Time Complexity: O(n)
    // Space complexity: O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 33 MB, less than 5.26%
    public int climbStairs3(int n) {
        if(n == 1) {
            return 1;
        }
        int[] mDP = new int[n+1];
        mDP[1] = 1;
        mDP[2] = 2;
        for(int i = 3; i <= n; i++) {
            mDP[i] = mDP[i - 1] + mDP[i-2];
        }
        return mDP[n];
    }
}

File: Golang_src/easy/climbingStairs_70.go
package easy

// Use DP:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 5.34%
func climbStairs(n int) int {
	if n == 1 {
		return 1
	}
	if n == 2 {
		return 2
	}
	// dp[i] = dp[i-1]+dp[i-2], for i >= 2
	// dp[0] = 1, dp[1] = 2
	dp := make([]int, n)
	dp[0], dp[1] = 1, 2

	for i := 2; i < n; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}

	return dp[n-1]
}

// Use Fib number:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 1.9 MB, less than 100.00%
func climbStairs2(n int) int {
	if n == 1 {
		return 1
	}
	first, second := 1, 2

	for i := 3; i <= n; i++ {
		third := first + second
		first = second
		second = third
	}

	return second
}
